pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,355 @@
+		return
+	}
+
+	f.logger.Logf("Login Successful. Current remote location : %q", f.config.RemoteDir)
+}
+
+// Create creates an empty file on the FTP server.
+func (f *ftpFileSystem) Create(name string) (container.File, error) {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Create", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Create_File failed. Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	emptyReader := new(bytes.Buffer)
+
+	err := f.conn.Stor(filePath, emptyReader)
+	if err != nil {
+		f.logger.Errorf("Create_File failed. Error creating file with path %q : %v", filePath, err)
+		return nil, err
+	}
+
+	_, s := path.Split(filePath)
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Create_File failed : %v", err)
+		return nil, err
+	}
+
+	f.logger.Logf("Create_File successful. Created file %s at %q", name, filePath)
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s,
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+		metrics:  f.metrics,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Mkdir", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Mkdir failed. Provide a valid directory : %v", errEmptyDirectory)
+		return errEmptyDirectory
+	}
+
+	err := f.conn.MakeDir(filePath)
+	if err != nil {
+		f.logger.Errorf("Mkdir failed. Error creating directory at %q : %v", filePath, err)
+		return err
+	}
+
+	f.logger.Logf("%s successfully created", name)
+
+	return nil
+}
+
+func (f *ftpFileSystem) mkdirAllHelper(filepath string) []string {
+	var dirs []string
+
+	currentdir := filepath
+
+	for {
+		err := f.conn.MakeDir(currentdir)
+		if err != nil {
+			parentDir, dir := path.Split(currentdir)
+
+			dirs = append([]string{dir}, dirs...)
+			if parentDir == "" || parentDir == "/" {
+				break
+			}
+			currentdir = path.Clean(parentDir)
+		} else {
+			dirs = append([]string{currentdir}, dirs...)
+			break
+		}
+
+	}
+
+	return dirs
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *ftpFileSystem) MkdirAll(name string, _ os.FileMode) error {
+	defer f.processLog(&FileLog{Operation: "MkdirAll", Location: path.Join(f.config.RemoteDir, name)}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("MkdirAll failed. Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+	// returns a slice of those directories that do not exist with the first index being the latest existing parent directory path.
+	dirs := f.mkdirAllHelper(name)
```

Comment:
I feel like you’re making your life harder than just using a recursive algorithm…

I will again recommend https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/os/path.go;l=18 as a model.

Reply:
I am just more comfortable with iterative algorithm. However it implements the same logic.